### DISCUSSION TITLE
Is there an "Invalidate" functionality for DI ?

### DISCUSSION THREAD

**Discussion by alexdonh** [TypedProject/ts-express-decorators#2030] [Q&A] [answered]
I got into a situation when I need to "invalidate" an already registered SINGLETON provider. In other words, I am looking for a way to replace a SINGLETON provider with a new instance. Is it possible? How can I achieve it? 

If I use sth like this:

```
@Service()
export class Service1 {

  @Inject(Something)
  protected something: Something;

  doThis() {
    if (new) {
       this.something = new Something(); // this does not work as it says TypeError: Cannot set property something of #<Service1> which has only a getter
    }
  }
}
```

@Romakita if you dont mind... ;)

**Comment by Romakita** [comment, marked as answer]
In this case, don’t use ˋInject` decorator. Use injectorService directly to rebuild the service. 

```typescript

@Service()
export class Service1 {

  @Inject()
  protected injector: InjectorService;
  protected something: Something;
 
   $onInit() {
     this.something = this.injector.invoke(Something);
   }
  doThis() {
    if (new) {
       this.something = this.injector.invoke(Something, undefined, { rebuild: true });
    }
  }
}

```

> Note: this example use internal DI Api and I won’t give warranty to maintain this in future. Use this at your own risk ;)

see you
Romain

**Reply by alexdonh** [reply]
I did have a look. it also set the new instance into injector service cache, meaning in other service if I use `Inject` decorator I can still get the new instance. Exactly what im looking for!!

Many thanks! 

PS: Personally I think this is something SINGLETON should have so pls dont deprecate it ;)

**Reply by Romakita** [reply]
Actually this option is used by PlatformTest. So I shouldn't be deprecated. The point is over his usage in a production code.
